### DISCUSSION TITLE
Nyquist and Audacity 3.2 realtime effects?

### DISCUSSION THREAD

**Discussion by werame** [audacity/audacity#2470] [Development]
I see from the git log there's a planned big new thing for Audacity 3.2 with real-time effects, which seem to approach it in most-DAWs style, with per track effects (but not Reaper-style with per-clip effects.) Anyhow...

What want to ask here (and I'm pinging @rbdannenberg and @Paul-Licameli): are there plans to make such effects writable in Nyquist? Right now the new RT effects seem to be C++ only as they have a new API not implemented by Nyquist.cpp. 

On a brainstorming-from-the-helicopter level, it seems to me that RT effects could be handled by Nyquist the way seq or trigger closures are presently handled.

**Comment by SteveDaulton** [comment]
Personally I'd love to see real-time Nyquist.
Some plug-ins would not be suitable for real-time, but I certainly think that it's possible to have real-time Nyquist effects, though it looks like a big job. Nyquist already has it's own implementation of sliders to control time varying signals, and many of Nyquist's primitives accept control signals as time varying parameters.
Perhaps this needs to be investigated alongside looking at the bigger picture of how (destructive) Nyquist effects interact with Audacity (such as how Nyquist accesses audio samples from Audacity).

**Reply by werame** [reply]
Well getting a "native Nyquist gui" would be interesting but not strictly necessary. For VSTs Audacity supports replacing the native plugin's gui with its own. I'm pretty sure this is done through VST automation. So if Nyquist had a similar facility, it would be relative simple to pass control signals from Audacity's gui to Nyquist. That would probalby be less involved than getting a foreign event loop to cooperate with wxWidgets, although the latter does have some support for that. 

SuperCollider for instance use OSC (albeit in fairly unique, idiosyncratic [way](https://opensoundcontrol.stanford.edu/publications/2004-Use-of-Open-Sound-Control-in-SuperCollider-Server.html)) to control its sound server's UGen controls. I think Nyquest has [some OSC support](https://ccrma.stanford.edu/groups/osc/implementations/Nyquist.html), but I'm not sure how easy it is to turn Sound object "knobs" via OSC in Nyquist. Perhaps @rbdannenberg can illuminate us on that... Actually I managed to find the [documentation on that](http://www.cs.cmu.edu/~rbd/doc/nyquist/part10.html#118). (The CCRMA paged linked to the wrong Nyquist sub-page on that.) Basically

> The only OSC message interpreted by Nyquist has an address of /slider, and two parameters: an integer slider number and a float value, nominally from 0.0 to 1.0.

Now the question is: can Audacity send OSC? 

Actually, that's probably not even necessary, as it could just call the underlying C functions and "plonk" the slider message.

**Reply by rbdannenberg** [reply]
There's a big "semantic gap" between VST and Nyquist. A big issue, even for something as simple as a filter, is that VST is synchronous audio-in and audio-out, so if you need 64 samples, you first go and shove 64 samples into the input, then you can get 64 samples out. In Nyquist, you go to the filter output and "ask" for 64 samples. The filter may or may not issue a callback (request) to provide any number of samples (up to some blocksize set at compile time). I you wanted to do something VST-like, you could pre-write a big input buffer with samples (or zero fill) and then run like VST: push 64 samples into the input, read 64 samples from the output. But to operate synchronously like that, you need a sizeable buffer which means you either have to use look-ahead to feed in samples from the "future" or the output is going to be delayed.

The "real time" features of Nyquist are pretty crude: There are SOUNDs that pull samples from "sliders." Here, a "slider" is really just a float stored in a global array of about 1000 "sliders". There are some GUI elements that can be connected to a slider (data actually gets from the NyquistIDE via stdin and special control characters) giving you a "real" slider, and you can set sliders via OSC. During lazy evaluation, when "slider sound" needs a block of samples, it goes to the "slider" and just makes N copies the *current* value into a block of size N and then Nyquist can manipulate the values as a signal.  If you run the slider sound at audio rate, you get about 40 requests/second, so the latency is about 25ms for the buffering, but then you probably need a smoothing filter that makes it more sluggish, and you have to add audio output buffer latency, which typically has to be about 200ms to avoid underruns during garbage collection, so realistically, latency is around 1/4 second. It's usable to demo synthesis algorithms with interactive control, and there's even a simple algorithmic composition demo with parametric controls and start/stop buttons all based on "slider sounds".

**Reply by werame** [reply]
Lookahead is usually not a problem for audio editors/DAWs, unlike for true live environments. I'm not being very original here, one of the Tracktion guys said it in a more creative way in a talk, something like "a DAW has perfect knowledge of the future"--I'm paraphrasing from memory. It's not quite true if you want to implement Ableton Live-like features where the DAW gets told by an external interface to start playing another clip in some out-of-order fashion while the rest of the track(s) are still playing as usual--the whole ["clip launching"](https://www.youtube.com/watch?v=EiwUN7hz6eU) affair, which some DAW devs like Reapers' have declined to [ever] implement themselves.

The whole plugin-delay-compensation business was invented for letting DAW plugins process data in delayed/lookahead buffers, e.g. dynamics compressor usually require it. You can get away without PDC if the effect is global so no desync can occur between tracks in the same project. 

By the way, the current GetCallback in Nyquist.cpp does some kind of buffer sizing:

```
      mCurBufferStart[ch] = (mCurStart[ch] + start);
      mCurBufferLen[ch] = mCurTrack[ch]->GetBestBlockSize(mCurBufferStart[ch]);

      if (mCurBufferLen[ch] < (size_t) len) {
         mCurBufferLen[ch] = mCurTrack[ch]->GetIdealBlockSize();
      }

      mCurBufferLen[ch] =
         limitSampleBufferSize( mCurBufferLen[ch],
                                mCurStart[ch] + mCurLen - mCurBufferStart[ch] );
```

I haven't followed those methods to see exactly how much they allocate.

**Reply by rbdannenberg** [reply]
My guess is block sizes here will be larger than typical Nyquist block sizes. Good point about lookahead in DAWs. In Nyquist, lookahead is unbounded, mostly because blocks of samples in low-sample-rate sounds span more time than blocks in audio-rate sounds, and because blocks are not necessarily aligned. But if you intend something for a realtime/interactive effect, you would want to run everything at audio rate, and everything should behave pretty synchronously, computing about 1000 samples at a time. You could either preload some buffers with effect input that Nyquist SOUNDs could pull from (as long as the effect does not try too much lookahead), or you could do something like we do now for track access, which is: wait for Nyquist to make a request, then go off and get the samples.

**Comment by Paul-Licameli** [comment]
> What want to ask here (and I'm pinging @rbdannenberg and @Paul-Licameli): are there plans to make such effects writable in Nyquist? Right now the new RT effects seem to be C++ only as they have a new API not implemented by Nyquist.cpp.

Fantasies, yes.

Plans, no.

**Reply by werame** [reply]
Mkay, so the future that Muse sees is just VST3? Is non-realtime support for that working in the VST3 PR? (For VST2 not many apps supported that non-realtime mode `VstOfflineTask`; Steinberg WaveLab is/was one. The "real" = Pro version only works with a USB dongle... so unusable on some laptops.)

N.B.  `git grep VstOfflineTask` returns nothing in the Audacity tree.

On the other hand, I see

```
   //defaults
   mSetup.processMode = Vst::kOffline;
```
in the VST3 PR. And Steinberg made the whole offline processing affair a lot less separate in VST3, I see. I'm not sure what is possible in VST3 offline mode mode though. In VST2 it [could do](https://www.kvraudio.com/forum/viewtopic.php?p=2362464#p2362464) a lot more than what online could, e.g. return a different length for a clip, something that Audacity also supports for its own C++ plugins (and Nyquist ones) in the non-realtime/offline mode.

**Reply by Paul-Licameli** [reply]
> Mkay, so the future that Muse sees is just VST3? 

I don't speak for Muse.  I speak for myself as one of what is now "the team."

Enhancing Nyquist would be great fun and great challenge, and I would love to see it done or have a hand in doing it, making Audacity an even better laboratory for the expert users who want to program it in a high level language.  But there so are many other initiatives occupying development efforts that we think will have higher value to more users.

Nyquist is, I hate to say, very peripheral to development priorites now.  On the other hand, I think that also means it is a part of the program where volunteer maintainers can be given a freer hand.

**Reply by LWinterberg** [reply]
> Mkay, so the future that Muse sees is just VST3?

in general, it's unhelpful to think of muse as this big monolithic thing that wants something. The muse C-suite has way too many other things to do over demanding individual feature decisions like this. 

> Nyquist is, I hate to say, very peripheral to development priorites now. On the other hand, I think that also means it is a part of the program where volunteer maintainers can be given a freer hand.

I would agree with this. @werame If you want to go nuts at improving nyquist support, we'll definitely support you in it. Although it may take some time to code review and merge in the changes you make, any improvements are much appreciated

**Comment by werame** [comment]
To end my digression on VST3: It might have been possible to shorten the output if [getTailSamples](https://steinbergmedia.github.io/vst3_doc/vstinterfaces/classSteinberg_1_1Vst_1_1IAudioProcessor.html#abb6b16a66b2356a4038a778b815dbbf3) were allowed to return a negative value, but it's not. So it seems that in VST3 an "offline" plugin effect has to return at least as much data as what's being pumped into it. I guess I can ask on SO or KVR if there's really no workaround for this.

I'm informed by a friend that VST being so limiting is among the reasons why e.g. Ableton has M4L (Max For Live) which [can](https://docs.cycling74.com/max5/refpages/m4l-ref/m4l_live_object_model.html) read/write entire tracks in the DAW etc.
